chaste_codegen/_lookup_tables.py

```diff
@@ -0,0 +1,207 @@
+                    self._set_lookup_table_if_appropriate(exp_func, vars_used, ex)
+                return False, vars_used_in_lut
+            return expensive_func, vars_used_in_lut
+
+    def _set_lookup_table_if_appropriate(self, exp_func, vars_used, expr):
+        """ Store an expression to the lookup table if it's suitable. """
+        # Expressions are suitable for lut if:
+        # - they have an expensive function
+        # - they're not already set as suitable
+        # - they only contain 1 variable and this variable is one of the lookup variables
+        if exp_func and expr not in self._lookup_table_expr and \
+                len(vars_used) == 1 and list(vars_used)[0] in self._lookup_variables:
+            self._lookup_table_expr[expr] = vars_used
+
+    def _process_lookup_parameters(self):
+        """ Prepare the stored lookup table parameters for generating chaste code.
+            *Please Note:* no more calls to `calc_lookup_tables` can be made after this. """
+        if not self._lookup_params_processed:
+            # Stick in a list of all expressions for the variable for easy access in the template
+            for param in self._lookup_parameters:
+                param[6] = list(filter(lambda k: param[5] in self._lookup_table_expr[k], self._lookup_table_expr))
+
+            # Filter out the parameter set for which we didn't find any complicated expressions
+            self._lookup_parameters = list(filter(lambda p: len(p[6]) > 0, self._lookup_parameters))
+            self._lookup_params_processed = True
+
+    def print_lut_expr(self, expr):
+        """ prints an individual lookup expression e.g. `lt_row_0[0].
+            *Please Note:* cannot be called after `print_lookup_parameters` has been called.
+            :param expr: the expression to print.
+            :return: a string with the printed lookup table expression if expr is in the lookup table
+                     and a method hass been associated, else None.
+
+            *Please Note:* in order to print lookup table expressions
+                           please associate a method_name via `method_being_printed`,
+                           if no method name is associated, None is returned."""
+        if self._lookup_params_printed:
+            raise ValueError('Cannot print lookup expression after main table has been printed')
+
+        self._process_lookup_parameters()
+        if self._method_printed and expr in self._lookup_table_expr:
+            variables = expr.free_symbols & self._lookup_variables
+            assert len(variables) == 1, "Lookup table expressions should have exactly 1 (lookup) variable"
+            var = tuple(variables)[0]
+            for i, param in enumerate(self._lookup_parameters):
+                if param[5] is var:
+                    param[4].add(self._method_printed)
+                    return '_lt_' + str(i) + '_row[' + str(param[6].index(expr)) + ']'
```

Comment:
Might be clearer to use .format?
```suggestion
                    return '_lt_{}_row[{}]'.format(i, param[6].index(expr))
```